homeassistant/components/binary_sensor/template.py

```diff
@@ -124,6 +144,12 @@ def template_bsensor_startup(event):
 
             self.hass.async_add_job(self.async_check_state)
 
+            if self._availability_template:
+                async_track_state_change(
```

Comment:
This should be incorporated with the previous state change listener and both templates should be processed in the same update method. The odds of these two templates being related is high.

Reply:
> This should be incorporated with the previous state change listener and both templates should be processed in the same update method. The odds of these two templates being related is high.

@balloob I split them because `_async_render` returns when the state is not available and so `async_check_state`, additionally `async_check_state` implements the delay logic. Instead I think availability should be considered regardless if the state template fails to render. So I added new methods because this was cleaner, while changing the current methods looked confusing.

But if you prefer I'll do the changes in the current methods without adding new ones.